### PR TITLE
CLN: handle EAs and fast path (no bounds checking) in safe_sort

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
This is a possible alternative solution to what we have been discussing in https://github.com/pandas-dev/pandas/pull/25592.

This moves the logic into `safe_sort`, with:

- adding a `check_outofbounds` keyword to disable extra checks (otherwise the performance benefit of `take_1d` is lost)
- fixing `safe_sort` to work for EAs

The `check_outofbounds` make it a bit more complicated, but without it, we can't benefit of the performance improvement for which `take_1d` was used originally in `factorize`.

(another solution is to simply decide that this performance improvement is not worth this extra code, and we simply use the current `safe_sort` (but fixed to work for EAs) in `factorize`)

Need to add some more tests for the combination of EAs with a custom na_sentinel (a case that is currently broken)

### CODE REVIEW

File: pandas/core/sorting.py [hunk 1]
```diff
@@ -425,6 +427,10 @@ def safe_sort(values, labels=None, na_sentinel=-1, assume_unique=False):
     assume_unique : bool, default False
         When True, ``values`` are assumed to be unique, which can speed up
         the calculation. Ignored when ``labels`` is None.
+    check_outofbounds : bool, default True
```
Comment (jreback) [inline pandas/core/sorting.py]: this is not a bad name but not consistent across pandas, we use verify elsewhere.

File: pandas/core/sorting.py [hunk 2]
```diff
@@ -494,15 +501,26 @@ def sort_mixed(values):
...
-    reverse_indexer.put(sorter, np.arange(len(sorter)))
-
-    mask = (labels < -len(values)) | (labels >= len(values)) | \
-        (labels == na_sentinel)
-
-    # (Out of bound indices will be masked with `na_sentinel` next, so we may
-    # deal with them here without performance loss using `mode='wrap'`.)
-    new_labels = reverse_indexer.take(labels, mode='wrap')
-    np.putmask(new_labels, mask, na_sentinel)
+    if na_sentinel == -1:
```
Comment (jreback) [inline pandas/core/sorting.py]: would rather just fix take_1d

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/sorting.py (hunk 1 above)
Comment (jreback) [reply pandas/core/sorting.py]: can you update & add a versionadded tag

File: pandas/core/sorting.py [hunk 3]
```diff
@@ -461,7 +467,8 @@ def sort_mixed(values):
         return np.concatenate([nums, np.asarray(strs, dtype=object)])
 
     sorter = None
-    if PY3 and lib.infer_dtype(values, skipna=False) == 'mixed-integer':
+    if (PY3 and not isinstance(values, ABCExtensionArray)
```
Comment (jreback) [inline pandas/core/sorting.py]: hah no more PY3 needed!

Comment (jreback) [inline pandas/core/sorting.py]: use is_extension_array

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
Coming back to this

> would rather just fix take_1d

For `take_1d` itself, I am not sure it makes a lot of sense to support this. Currently, it is by definition that -1 values in the `indexer` passed to take signal missing values. The only use case for having this configurable would be this one, while it complicates the interface throughout pandas (eg also EAs would need to support this, as `take_1d` dispatch to them)

**Comment by jreback** [comment]
can you merge master

**Comment by jorisvandenbossche** [comment]
@jreback conflicts resolved, if you can take another look

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jorisvandenbossche nice cleanup & tests.

**Pull request merged**